src/utils/daos_autotest.c

```diff
@@ -0,0 +1,852 @@
+		if (rc) {
+			step_fail("failed to open object: %s", d_errdesc(rc));
+			return -1;
+		}
+
+		rc = daos_obj_close(oh, NULL);
+		if (rc) {
+			step_fail("failed to close object: %s", d_errdesc(rc));
+			return -1;
+		}
+	}
+
+	step_success("");
+	return 0;
+}
+
+static int
+kv_put(daos_handle_t oh, daos_size_t size, uint64_t nr)
+{
+	daos_handle_t	eq;
+	daos_event_t	ev_array[MAX_INFLIGHT];
+	char		key[MAX_INFLIGHT][10];
+	char		*val;
+	daos_event_t	*evp;
+	uint64_t	i;
+	int		rc;
+	int		ret;
+
+	/** Create event queue to manage asynchronous I/Os */
+	rc = daos_eq_create(&eq);
+	if (rc) {
+		return rc;
+	}
+
+	/** allocate buffer to store value */
+	val = malloc(size * MAX_INFLIGHT);
+	if (val == NULL) {
+		rc = daos_eq_destroy(eq, 0);
+		return -DER_NOMEM;
+	}
+	memset(val, 'D', size * MAX_INFLIGHT);
+
+	/** Issue actual I/Os */
+	for (i = 1; i < nr + 1; i++) {
+		char *key_cur;
+		char *val_cur;
+
+		if (i < MAX_INFLIGHT) {
+			/** Haven't reached max request in flight yet */
+			evp = &ev_array[i];
+			rc = daos_event_init(evp, eq, NULL);
+			if (rc) {
+				break;
+			}
+			key_cur = key[i];
+			val_cur = val + size * i;
+		} else {
+			int slot;
+
+			/**
+			 * Max request in flight reached, wait for one i/o to
+			 * complete to reuse the slot
+			 */
+			rc = daos_eq_poll(eq, 1, DAOS_EQ_WAIT, 1, &evp);
+			if (rc < 0)
+				break;
+			if (rc == 0) {
+				rc = -DER_IO;
+				break;
+			}
+
+			/** Check if completed operation failed */
+			if (evp->ev_error != DER_SUCCESS) {
+				rc = evp->ev_error;
+				break;
+			}
+			evp->ev_error = 0;
+
+			/* reuse slot */
+			slot = evp - ev_array;
+			key_cur = key[slot];
+			val_cur = val + slot * size;
+		}
+
+		/** key = insert sequence ID */
+		sprintf(key_cur, "%ld", i);
+		/** value = sequend ID + DDDDDDD... */
+		*((uint64_t *)val_cur) = i;
+
+		/** Insert kv pair */
+		rc = daos_kv_put(oh, DAOS_TX_NONE, 0, key_cur, size, val_cur,
+				evp);
+		if (rc) {
+			break;
+		}
+	}
+
+	/** Wait for completion of all in-flight requests */
+	do {
+		ret = daos_eq_poll(eq, 1, DAOS_EQ_WAIT, 1, &evp);
+		if (rc == 0 && ret == 1) {
+			rc = evp->ev_error;
+		}
+	} while (ret == 1);
+
+	if (rc == 0 && ret < 0) {
+		rc = ret;
+	}
+
+	free(val);
+
+	/** Destroy event queue */
+	ret = daos_eq_destroy(eq, 0);
+	if (ret) {
+		if (rc == 0)
+			rc = ret;
+	}
+
+	return rc;
+}
+
+static int
+kv_get(daos_handle_t oh, daos_size_t size, uint64_t nr)
+{
+	daos_handle_t	eq;
+	daos_event_t	ev_array[MAX_INFLIGHT];
+	char		key[MAX_INFLIGHT][10];
+	daos_size_t	val_sz[MAX_INFLIGHT];
+	char		*val;
+	daos_event_t	*evp;
+	uint64_t	i;
+	uint64_t	res = 0;
+	int		rc;
+	int		ret;
+
+	/** Create event queue to manage asynchronous I/Os */
+	rc = daos_eq_create(&eq);
+	if (rc) {
+		return rc;
+	}
+
+	/** allocate buffer to store value */
+	val = malloc(size * MAX_INFLIGHT);
+	if (val == NULL) {
+		ret = daos_eq_destroy(eq, 0);
+		return -DER_NOMEM;
+	}
+
+	/** Issue actual I/Os */
+	for (i = 1; i < nr + 1; i++) {
```

Comment:
This index is actually stored in the key and value. In the read phase, i am verifying that I got the right data back by summing all the values that i have seen. I should thus get SUM(i){1..nr} = nr * (nr + 1) / 2. I did not want to add any confusion for the reader by starting at 0, so i decided to start at 1. Clear?

Reply:
Sure. It just threw me a little bit since we're using it to index into arrays. Guess it doesn't hurt anything in this case since the arrays are just temporary storage.